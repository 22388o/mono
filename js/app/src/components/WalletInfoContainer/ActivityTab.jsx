import { useCallback, useState, useSyncExternalStore } from "react";
import { Button, Divider, Stack, Typography } from "@mui/material"
import ArrowOutwardIcon from '@mui/icons-material/ArrowOutward';

import styles from "../../styles/WalletInfoContainer.module.css";
import { SWAP_STATUS, getStringFromDate } from "../../utils/helpers";
import { walletStore } from "../../syncstore/walletstore";
import { activitiesStore } from "../../syncstore/activitiesstore";

export const ActivityTab = () => {
  const activities = useSyncExternalStore(activitiesStore.subscribe, () => activitiesStore.currentState);

  /**
   * Handler to cancel swap
   */
  const onCancelSwap = useCallback((e, activity) => {
    e.stopPropagation();
    if(activity.baseAsset === 'BTC') walletStore.dispatch({ type: 'ADD_NODE_BALANCE', payload: activity.baseQuantity });
    else if(activity.baseAsset === 'ETH') walletStore.dispatch({ type: 'ADD_WALLET_BALANCE', payload: activity.baseQuantity });
    console.log(activity, activities);
    activitiesStore.dispatch({ type: 'CANCEL_SWAP', payload: { secretHash: activity.hash } });
  }, [walletStore, activities]);

  /**
   * Component Render Function
   * @param {object} activity 
   * @returns Component for activity
   */
  const renderActivity = (activity) => {
    return (
      <Stack className={`${styles['activity-item']} activity-item panelActivity`} direction='row'>
        <Stack direction='row' gap={2}>
          <ArrowOutwardIcon sx={{color:'#6A6A6A', borderRadius: '50%', backgroundColor:'#101010', fontSize: '15px', padding: '5px' }} />
          <Stack sx={{alignItems:'flex-start'}}>
            <Typography>{ SWAP_STATUS[activity.status] }</Typography>
            <Typography sx={{fontSize: '14px', color: '#6A6A6A'}}>{ getStringFromDate(activity.createdDate) }</Typography>
            {/* <Typography sx={{fontSize:'12px'}}>{ activity.status === 5 && activity.hash.slice(0, 20) + '...'  }</Typography> */}
          </Stack>
        </Stack>
        <Stack>
          <Stack direction='row' gap={1}>
            <Typography>- {activity.baseQuantity.toFixed(5).replace(/[.,]0+$/ , "")}</Typography>
            <Typography sx={{color: '#6A6A6A'}}>{activity.baseAsset}</Typography>
          </Stack>
          <Stack direction='row' gap={1}>
            <Typography sx={{color: '#6A6A6A'}}>+ {activity.quoteQuantity.toFixed(5).replace(/[.,]0+$/ , "")}</Typography>
            <Typography sx={{color: '#6A6A6A'}}>{activity.quoteAsset}</Typography>
          </Stack>
          { activity.status < 11 && <span><Button onClick={(e) => onCancelSwap(e, activity)} className={`${styles['cancel-btn']} cancel-btn`}>
            Cancel
          </Button></span> }
        </Stack>
      </Stack>
    )
  }

  /**
   * Divider Render
   */
  const renderDivider = () => <Divider sx={{borderColor: '#2A2A2A!important', margin: '15px 0px'}} />

  return (
    <Stack className={styles['activities-container']}>
      { [...activities].reverse().map((activity, index) =>
          <div key={index}>
            {index > 0 && renderDivider() }
            { renderActivity(activity) }
          </div>
        )
      }

      { activities.length === 0 && <Typography sx={{ color: '#555555', fontSize: '20px', marginTop: '100px' }}>No Activities</Typography>}
    </Stack>
  )
}
