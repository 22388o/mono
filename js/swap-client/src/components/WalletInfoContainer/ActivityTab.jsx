import { useCallback, useState, useSyncExternalStore } from "react";
import { Button, Divider, Stack, Typography } from "@mui/material"
import ArrowOutwardIcon from '@mui/icons-material/ArrowOutward';

import styles from "../../styles/WalletInfoContainer.module.css";
import { IndexedDB } from "@portaldefi/sdk";
import { SWAP_STATUS, getStringFromDate } from "../../utils/helpers";
import { IndexedDB_dispatch } from "../../utils/indexeddb";
import { walletStore } from "../../syncstore/walletstore";

export const ActivityTab = () => {
  const activities = useSyncExternalStore(IndexedDB.subscribe, IndexedDB.getAllActivities);

  const onCancelSwap = useCallback((e, activity, index) => {
    e.stopPropagation();
    IndexedDB_dispatch({ type: 'CANCEL_SWAP', payload: index });
    if(activity.baseAsset === 'BTC') walletStore.dispatch({ type: 'ADD_NODE_BALANCE', payload: activity.baseQuantity });
    else if(activity.baseAsset === 'ETH') walletStore.dispatch({ type: 'ADD_WALLET_BALANCE', payload: activity.baseQuantity });
    else walletStore.dispatch({ type: 'ADD_NFT_BALANCE', payload: {type: activity.baseAsset, balance: 1} });
  }, [walletStore, IndexedDB]);

  const renderActivity = (activity, index) => {
    return (
      <Stack className={styles['activity-item']} direction='row'>
        <Stack direction='row' gap={2}>
          <ArrowOutwardIcon sx={{color:'#6A6A6A', borderRadius: '50%', backgroundColor:'#101010', fontSize: '15px', padding: '5px' }} />
          <Stack sx={{alignItems:'flex-start'}}>
            <Typography>{ SWAP_STATUS[activity.status] }</Typography>
            <Typography sx={{fontSize: '14px', color: '#6A6A6A'}}>{ getStringFromDate(activity.createdDate) }</Typography>
            <Typography sx={{fontSize:'12px'}}>{ activity.status === 5 && activity.hash.slice(0, 20) + '...'  }</Typography>
          </Stack>
        </Stack>
        <Stack>
          <Stack direction='row' gap={1}>
            <Typography>- {activity.baseQuantity.toFixed(5).replace(/[.,]0+$/ , "")}</Typography>
            <Typography sx={{color: '#6A6A6A'}}>{activity.baseAsset}</Typography>
          </Stack>
          <Stack direction='row' gap={1}>
            <Typography sx={{color: '#6A6A6A'}}>+ {activity.quoteQuantity.toFixed(5).replace(/[.,]0+$/ , "")}</Typography>
            <Typography sx={{color: '#6A6A6A'}}>{activity.quoteAsset}</Typography>
          </Stack>
          { activity.status < 5 && <span><Button onClick={(e) => onCancelSwap(e, activity, index)} className={styles['cancel-btn']}>
            Cancel
          </Button></span> }
        </Stack>
      </Stack>
    )
  }

  const renderDivider = () => <Divider sx={{borderColor: '#2A2A2A!important', margin: '15px 0px'}} />

  return (
    <Stack className={styles['activities-container']}>
      { [...activities].reverse().map((activity, index) =>
          <div>
            {index > 0 && renderDivider() }
            { renderActivity(activity, index) }
          </div>
        )
      }

      { activities.length === 0 && <Typography sx={{ color: '#555555', fontSize: '20px', marginTop: '100px' }}>No Activities</Typography>}
    </Stack>
  )
}