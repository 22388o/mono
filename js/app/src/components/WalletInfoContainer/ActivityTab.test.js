import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom'; // For the "toBeInTheDocument" matcher
import { ActivityTab } from './ActivityTab';

describe('ActivityTab component test', () => {
  it('renders component', async () => {

    const {container} = render(
      <ActivityTab />
    );

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
