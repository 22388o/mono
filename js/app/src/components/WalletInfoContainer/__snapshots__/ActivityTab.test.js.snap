// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityTab component test renders component 1`] = `
<div>
  <div
    class="MuiStack-root css-nen11g-MuiStack-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-1603n5b-MuiTypography-root"
    >
      No Activities
    </p>
  </div>
</div>
`;
